lib/stdlib/src/peer.erl

```diff
@@ -370,38 +412,101 @@ handle_info({tcp_closed, Sock}, #peer_state{connection = Sock} = State) ->
+        {{halt, Timeout}, error} ->
+            try
+                _ = erpc:call(Node, erlang, halt, [], Timeout),
+                ok
+            catch
+                error:{erpc,noconnection} -> ok;
+                _:_ -> force_disconnect_node(Node)
+            end;
         {Shutdown, error} ->
             Timeout = shutdown(dist, undefined, Node, Shutdown),
-            receive {nodedown, Node} -> ok after Timeout -> ok end;
+            wait_disconnected(Node, {timeout, Timeout});
         {Shutdown, {ok, standard_io}} ->
             Timeout = shutdown(port, Port, Node, Shutdown),
+            Deadline = deadline(Timeout),
             receive {'EXIT', Port, _Reason2} -> ok after Timeout -> ok end,
-            catch erlang:port_close(Port);
+            catch erlang:port_close(Port),
+            wait_disconnected(Node, Deadline);
         {Shutdown, {ok, _TCP}} ->
             Timeout = shutdown(tcp, Port, Node, Shutdown),
+            Deadline = deadline(Timeout),
             receive {tcp_closed, Port} -> ok after Timeout -> ok end,
-            catch catch gen_tcp:close(Port)
+            catch catch gen_tcp:close(Port),
+            wait_disconnected(Node, Deadline)
     end,
     ok.
 
 %%--------------------------------------------------------------------
 %% Internal implementation
 
+deadline(infinity) ->
+    {timeout, infinity};
+deadline(Timeout) when is_integer(Timeout) ->
+    {deadline, erlang:monotonic_time(millisecond) + Timeout}.
+
+wait_disconnected(Node, WaitUntil) ->
+    %% Should only be called just before we are exiting the caller, so
+    %% we do not bother disabling nodes monitoring if we enable it and
+    %% do not flush any nodeup/nodedown messages that we got due to the
+    %% nodes monitoring...
+    case lists:member(Node, nodes(connected)) of
+        false ->
+            ok;
+        true ->
+            _ = net_kernel:monitor_nodes(true, [{node_type, all}]),
+            %% Need to check connected nodes list again, since it
+            %% might have disconnected before we enabled nodes
+            %% monitoring...
+            case lists:member(Node, nodes(connected)) of
+                false ->
+                    ok;
+                true ->
+                    Tmo = case WaitUntil of
+                              {timeout, T} ->
+                                  T;
+                              {deadline, T} ->
+                                  TL = T - erlang:monotonic_time(millisecond),
+                                  if TL < 0 -> 0;
+                                     true -> TL
+                                  end
+                          end,
+                    receive {nodedown, Node, _} -> ok
+                    after Tmo -> force_disconnect_node(Node)
+                    end
+            end
+    end.
+
+force_disconnect_node(Node) ->
+    _ = erlang:disconnect_node(Node),
+    error_logger:warning_msg("peer:stop() timed out waiting for disconnect from "
```

Comment:
For my education, is `error_logger` still recommended to use? With new `logger` and `?LOG_WARNING` macro available, I'd think of using those instead of `error_logger` (which I thought was deprecated).

Also, it'd be great to have some guidelines on logging within OTP itself. I'd think of adding `otp` domain to all messages emitted within OTP itself (for easier filtering).

BTW is it necessary to log this event?

Reply:
> For my education, is `error_logger` still recommended to use? With new `logger` and `?LOG_WARNING` macro available, I'd think of using those instead of `error_logger` (which I thought was deprecated).
> 

I guess `logger` is supposed to be used. Fixed it.

> Also, it'd be great to have some guidelines on logging within OTP itself. I'd think of adding `otp` domain to all messages emitted within OTP itself (for easier filtering).
> 

I'm not the right person to answer this. @garazdawi can probably answer this when he is back (currently out of office).

> BTW is it necessary to log this event?

Yes I *really* think so. It is really annoying trying to find the reason to why a test case failed when there is no information about what happened. This information is also only logged when something has gone wrong and should normally never happen.

BTW regarding logging. I really miss the old style `test_server:start_node()` logs where you could see in the test log that the node had come up and how it was started with the exact command line arguments. Those logs have helped me numerous times in the past.